homeassistant/components/netgear/router.py

```diff
@@ -0,0 +1,212 @@
+
+        self.listeners = []
+        self._unsub_dispatcher = None
+
+    def _setup(self) -> None:
+        """Set up a Netgear router sync portion."""
+        self._api = get_api(
+            self._password,
+            self._host,
+            self._username,
+            self._port,
+            self._ssl,
+            self._url,
+        )
+
+        self._info = self._api.get_info()
+        self.device_name = self._info["DeviceName"]
+        self.model = self._info["ModelName"]
+        self.firmware_version = self._info["Firmwareversion"]
+
+        if self.model in MODELS_V2:
+            self._method_version = 2
+
+    async def async_setup(self) -> None:
+        """Set up a Netgear router."""
+        await self.hass.async_add_executor_job(self._setup)
+
+        # set already known devices to away instead of unavailable
+        entity_registry = er.async_get(self.hass)
+        entries = er.async_entries_for_config_entry(entity_registry, self.entry_id)
+        for entity_entry in entries:
+            self.devices[entity_entry.unique_id] = {
+                "mac": entity_entry.unique_id,
+                "name": entity_entry.original_name,
+                "active": False,
+                "last_seen": dt_util.utcnow() - timedelta(days=365),
+                "device_model": None,
+                "device_type": None,
+                "type": None,
+                "link_rate": None,
+                "signal": None,
+                "ip": None,
+            }
+
+        await self.async_update_device_trackers()
+        self._unsub_dispatcher = async_track_time_interval(
+            self.hass, self.async_update_device_trackers, SCAN_INTERVAL
+        )
+
+        async_dispatcher_send(self.hass, self.signal_device_new)
+
+    async def async_unload(self) -> None:
+        """Unload a Netgear router."""
+        self._unsub_dispatcher()
+        self._unsub_dispatcher = None
+
+    async def async_get_attached_devices(self) -> None:
+        """Get the devices connected to the router."""
+        if self._method_version == 1:
+            return await self.hass.async_add_executor_job(
+                self._api.get_attached_devices
+            )
+
+        return await self.hass.async_add_executor_job(self._api.get_attached_devices_2)
+
+    async def async_update_device_trackers(self, now=None) -> None:
+        """Update Netgear devices."""
+        new_device = False
+        ntg_devices = await self.async_get_attached_devices()
+        now = dt_util.utcnow()
+
+        for ntg_device in ntg_devices:
+            device_mac = format_mac(ntg_device.mac)
+
+            if self._method_version == 2 and not ntg_device.link_rate:
+                continue
+
+            if self._tracked_list and device_mac not in self._tracked_list:
+                continue
+
+            if not self.devices.get(device_mac):
+                new_device = True
+
+            self.devices[device_mac] = ntg_device._asdict()
```

Comment:
Can we avoid accessing the protected function here?

Reply:
ntg_device is a namedtuple from the collections module: `from collections import namedtuple`.
I think _asdict() is the proper way to convert it to a dictionary and I don't know of another way....
Do you have sugestions?

Reply:
Looks like namedtuple previously changed what `_asdict` returned in 3.8 (probably doesn't matter in this case) https://docs.python.org/3/library/collections.html#collections.somenamedtuple._asdict

Maybe use a dataclass instead as its api seems to have better stability guarantee:
https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

Reply:
I don't maintain the upstream lib, so I preffer to leave it like this since I don't know what applications are using the upstream lib. (the Namedtuple is made in the upstream lib).

Reply:
Maybe add a comment explaining the use of the protected method

Reply:
I did